WordPress/gutenberg
# Fix block toolbar positioning

## What?
Attempts to fix #41575, where the block toolbar doesn't flip position if there's no space at the top of the editor canvas.

## How?
I've tried introducing a new prop `__unstableAvoidOverflow` that when set runs new floating-ui middleware.

For Popovers that have a 'top' placement, this middleware detects if there's enough space for the popover in that top placement. If there isn't, it flips the popover to be positioned at the bottom using the `flip` middleware.

When there is space at the top of the , the middleware falls back to calling the `shift` middleware. This makes it so that when a larger block is selected, scrolling will cause the toolbar to adopt a sticky behavior at the top of the screen. It isn't possible to use the `__unstableShift` prop for this second part because when that's active it causes the first 'flip' behavior I described to have no effect (I'm unsure if this is a floating-ui bug).

## Testing Instructions
Testing that the toolbar flips when there's no space at the top of the canvas
1. Use empty theme
2. Visit the site editor
3. Select the top-most block

Expected - the toolbar should be under the block

Testing that the toolbar shifts over the block when scrolling (this is not currently working)
1. Open the demo post
2. Select the cover block
3. Scroll down

Expected - the toolbar should maintain a sticky position at the top of the viewport.

## Screenshots or screencast 
### Before
![Screen Shot 2022-07-13 at 5 15 16 pm](https://user-images.githubusercontent.com/677833/178697731-4396e495-57e0-4c5e-9aa4-d6e9572dea4c.png)

### After

https://user-images.githubusercontent.com/677833/178697407-22b944e8-2ee1-4948-be39-72b2b55660c3.mp4
